GANDLF/models/lightning_module.py

```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+        memory_stats_dir = self._prepare_memory_stats_save_dir()
+        full_filepath = os.path.join(memory_stats_dir, filename)
+        file_write_mode = "a" if os.path.exists(full_filepath) else "w"
+        using_cuda = "cuda" in self.device.type
+
+        memory_info_string = "Epoch,Memory_Total,Memory_Available,Memory_Percent_Free,Memory_Usage,"  # used to write output
+        if using_cuda:
+            memory_info_string += (
+                "CUDA_active.all.peak,CUDA_active.all.current,CUDA_active.all.allocated"
+            )
+        memory_info_string += "\n"
+
+        host_memory_stats = psutil.virtual_memory()
+        memory_info_string += (
+            str(epoch)
+            + ","
+            + str(host_memory_stats[0])
+            + ","
+            + str(host_memory_stats[1])
+            + ","
+            + str(host_memory_stats[2])
+            + ","
+            + str(host_memory_stats[3])
+        )
+        if using_cuda:
+            cuda_memory_stats = torch.cuda.memory_stats()
+            memory_info_string += (
+                ","
+                + str(cuda_memory_stats["active.all.peak"])
+                + ","
+                + str(cuda_memory_stats["active.all.current"])
+                + ","
+                + str(cuda_memory_stats["active.all.allocated"])
+            )
+        memory_info_string += ",\n"
+        with open(full_filepath, file_write_mode) as file_mem:
+            file_mem.write(memory_info_string)
+
+    @rank_zero_only
+    def _prepare_memory_stats_save_dir(self):
+        memory_stats_dir = os.path.join(self.output_dir, "memory_stats")
+        os.makedirs(memory_stats_dir, exist_ok=True)
+        return memory_stats_dir
+
+    @rank_zero_only
+    def _print_epoch_start_time(self):
+        print("Epoch start time : ", get_date_time(), flush=True)
+
+    @rank_zero_only
+    def _set_epoch_start_time(self):
+        self.epoch_start_time = time.time()
+
+    # TODO when used with multple GPUs, thil will produce multiple logs
+    # for each GPU. We should think on doing allgather here in a function
+    # that is called on the main process (rank 0)
+
+    def on_train_epoch_end(self):
+        training_epoch_average_metrics = {}
+        metric_names = self.training_metric_values[0].keys()
+        for metric_name in metric_names:
+            metric_values = [x[metric_name] for x in self.training_metric_values]
+            training_epoch_average_metrics[metric_name] = mean(metric_values)
+
+        train_losses_gathered = self.all_gather(self.train_losses)
+        mean_loss = torch.mean(torch.stack(train_losses_gathered)).item()
+        # Note that we DO NOT have overall stats calculators here as in the original code
+        # Is it valid? Should we add it? This woudl require accumulation of all the predictions and labels
+        # thus greatly increasing memory usage, especially in segmentation tasks
```

Comment:
So, the `overall_stats` are only calculated for classification and regression tasks, and hence doesn't have any issues WRT memory usage. I feel we need to have this on.

Reply:
Right, somehow did not comprehend the conditional statement. Will update this one